backend/src/nodes/impl/dithering/constants.py

```diff
@@ -0,0 +1,209 @@
+        (1, -1): 1 / 8,
+        (1, 0): 1 / 8,
+        (1, 1): 1 / 8,
+        (2, 0): 1 / 8,
+    },
+    ErrorDiffusionMap.BURKES: {
+        (0, 1): 8 / 32,
+        (0, 2): 4 / 32,
+        (1, -2): 2 / 32,
+        (1, -1): 4 / 32,
+        (1, 0): 8 / 32,
+        (1, 1): 4 / 32,
+        (1, 2): 2 / 32,
+    },
+    ErrorDiffusionMap.FLOYD_STEINBERG: {
+        (0, 1): 7 / 16,
+        (1, -1): 3 / 16,
+        (1, 0): 5 / 16,
+        (1, 1): 1 / 16,
+    },
+    ErrorDiffusionMap.JARVIS_ET_AL: {
+        (0, 1): 7 / 48,
+        (0, 2): 5 / 48,
+        (1, -2): 3 / 48,
+        (1, -1): 5 / 48,
+        (1, 0): 7 / 48,
+        (1, 1): 5 / 48,
+        (1, 2): 3 / 48,
+        (2, -2): 1 / 48,
+        (2, -1): 3 / 48,
+        (2, 0): 5 / 48,
+        (2, 1): 3 / 48,
+        (2, 2): 1 / 48,
+    },
+    ErrorDiffusionMap.SIERRA: {
+        (0, 1): 5 / 32,
+        (0, 2): 3 / 32,
+        (1, -2): 2 / 32,
+        (1, -1): 4 / 32,
+        (1, 0): 5 / 32,
+        (1, 1): 4 / 32,
+        (1, 2): 2 / 32,
+        (2, -1): 2 / 32,
+        (2, 0): 3 / 32,
+        (2, 1): 2 / 32,
+    },
+    ErrorDiffusionMap.TWO_ROW_SIERRA: {
+        (0, 1): 4 / 16,
+        (0, 2): 3 / 16,
+        (1, -2): 1 / 16,
+        (1, -1): 2 / 16,
+        (1, 0): 3 / 16,
+        (1, 1): 2 / 16,
+        (1, 2): 1 / 16,
+    },
+    ErrorDiffusionMap.SIERRA_LITE: {(0, 1): 2 / 4, (1, -1): 1 / 4, (1, 0): 1 / 4},
+    ErrorDiffusionMap.STUCKI: {
+        (0, 1): 8 / 42,
+        (0, 2): 4 / 42,
+        (1, -2): 2 / 42,
+        (1, -1): 4 / 42,
+        (1, 0): 8 / 42,
+        (1, 1): 4 / 42,
+        (1, 2): 2 / 42,
+        (2, -2): 1 / 42,
+        (2, -1): 2 / 42,
+        (2, 0): 4 / 42,
+        (2, 1): 2 / 42,
+        (2, 2): 1 / 42,
+    },
+}
+
+# Sanity check
+for error_diffusion_map, coords in ERROR_DIFFUSION_MAPS.items():
+    for (row, column) in coords.keys():
+        if row < 0 or (row == 0 and column <= 0):
+            logger.warn(
+                f"Error diffusion map {error_diffusion_map} has an illegal coordinate: {row}, {column}"
+            )
+
+
+# https://en.wikipedia.org/wiki/Ordered_dithering
+
+
+class ThresholdMap(Enum):
+    BAYER_2 = "B2"
+    BAYER_4 = "B4"
+    BAYER_8 = "B8"
+    BAYER_16 = "B16"
+
+
+THRESHOLD_MAP_LABELS = {
+    ThresholdMap.BAYER_2: "Bayer 2x2",
+    ThresholdMap.BAYER_4: "Bayer 4x4",
+    ThresholdMap.BAYER_8: "Bayer 8x8",
+    ThresholdMap.BAYER_16: "Bayer 16x16",
+}
+THRESHOLD_MAPS = {
+    ThresholdMap.BAYER_2: np.array(
+        list(
+            map(
+                int,
+                """
+        0 2
+        3 1
+        """.strip().split(),
+            )
+        )
+    ).reshape((2, 2)),
+    ThresholdMap.BAYER_4: np.array(
+        list(
+            map(
+                int,
+                """
+         0  8  2 10
+        12  4 14  6
+         3 11  1  9
+        15  7 13  5
+        """.strip().split(),
+            )
+        )
+    ).reshape((4, 4)),
```

Comment:
You did this to get around black messing up the formatting, right? I think a better solution might be to either turn off formatting (`# fmt: off`) for those lines or to use [some of the tricks in this post](https://stackoverflow.com/questions/58584413/black-formatter-ignore-specific-multi-line-code)

Reply:
Nice.  Yes, this looks a lot better.